contracts/meta/MetaSwap.sol

```diff
@@ -0,0 +1,426 @@
+     * @param _a the amplification coefficient * n * (n - 1). See the
+     * StableSwap paper for details
+     * @param _fee default swap fee to be initialized with
+     * @param _adminFee default adminFee to be initialized with
+     * @param _withdrawFee default withdrawFee to be initialized with
+     */
+    function initialize(
+        IERC20[] memory _pooledTokens,
+        uint8[] memory decimals,
+        string memory lpTokenName,
+        string memory lpTokenSymbol,
+        uint256 _a,
+        uint256 _fee,
+        uint256 _adminFee,
+        uint256 _withdrawFee
+    ) public virtual override initializer {
+        revert("use initializeMetaSwap() instead");
+    }
+
+    /**
+     * @notice Initializes this MetaSwap contract with the given parameters.
+     * MetaSwap uses an existing Swap pool to expand the available liquidity.
+     * _pooledTokens array should contain the base Swap pool's LP token as
+     * the last element. For example, if there is a Swap pool consisting of
+     * [DAI, USDC, USDT]. Then a MetaSwap pool can be created with [sUSD, BaseSwapLPToken]
+     * as _pooledTokens.
+     *
+     * This will also deploy the LPToken that represents users
+     * LP position. The owner of LPToken will be this contract - which means
+     * only this contract is allowed to mint new tokens.
+     *
+     * @param _pooledTokens an array of ERC20s this pool will accept. The last
+     * element must be an existing Swap pool's LP token's address.
+     * @param decimals the decimals to use for each pooled token,
+     * eg 8 for WBTC. Cannot be larger than POOL_PRECISION_DECIMALS
+     * @param lpTokenName the long-form name of the token to be deployed
+     * @param lpTokenSymbol the short symbol for the token to be deployed
+     * @param _a the amplification coefficient * n * (n - 1). See the
+     * StableSwap paper for details
+     * @param _fee default swap fee to be initialized with
+     * @param _adminFee default adminFee to be initialized with
+     * @param _withdrawFee default withdrawFee to be initialized with
+     */
+    function initializeMetaSwap(
+        IERC20[] memory _pooledTokens,
+        uint8[] memory decimals,
+        string memory lpTokenName,
+        string memory lpTokenSymbol,
+        uint256 _a,
+        uint256 _fee,
+        uint256 _adminFee,
+        uint256 _withdrawFee,
+        ISwap baseSwap
+    ) public virtual initializer {
+        Swap.initialize(
+            _pooledTokens,
+            decimals,
+            lpTokenName,
+            lpTokenSymbol,
+            _a,
+            _fee,
+            _adminFee,
+            _withdrawFee
+        );
+
+        // MetaSwap initializer
+        metaSwapStorage.baseSwap = baseSwap;
+        metaSwapStorage.baseVirtualPrice = baseSwap.getVirtualPrice();
+        metaSwapStorage.baseCacheLastUpdated = block.timestamp;
+        for (uint8 i; i < 32; i++) {
+            try baseSwap.getToken(i) returns (IERC20 token) {
```

Comment:
What if this doesn't complete for some reason?

Reply:
Added additional check for baseSwap to have at least 2 tokens